### PR TITLE
Added latency support for devices + logging reasons why devices marked as down

### PR DESCRIPTION
**Pull request by laf**
Yes I know this is two PRs in one really. I've done it as two commits in one so can separate out if we need.

#1307 is dealt with now on this. We log packet loss, min, max and avg latency to the db (cleared out in daily.sh) so that we can generate graphs like this:

image

You can then also use macros to alert on % packet loss over 15 mins or 5 mins.

Second commit deals with #1305. Now logs the last reason for a device changing status (icmp / snmp or back to blank if it goes up). In the devices list instead of saying down it will say icmp or snmp so you can visually see why. Also means you can include this in an alert template.

### CODE REVIEW

File: includes/functions.php [hunk 1]
```diff
@@ -495,25 +495,21 @@ function isPingable($hostname,$device_id = FALSE)
...
+    if(is_numeric($config['fping_options']['count']) || $config['fping_options']['count'] > 0) {
+        $fping_params .= ' -c ' . $config['fping_options']['count'];
+    }
+   //$status = shell_exec($config['fping'] . "$fping_params -e $hostname 2>/dev/null");
    $response = array();
-   if (strstr($status, "alive"))
-   {
-     $response['result'] = TRUE;
+   $status = fping($hostname,$fping_params);
+   if ($status['loss'] < 0 || $status['loss'] == 100) {
```
Comment (f0o) [inline includes/functions.php]: why `== 100` ?

Comment (laf) [inline includes/functions.php]: That's for detecting it's down, so if a device has 100% packet loss then it's down (if no value is returned this is due to a bad hostname so we class that as down).

my https://librenms.lathwood.co.uk branch is running this at present if you want a mix of up / down devices.

**Comment by laf** [comment]
Scrut is now duplicate code + 1 unused which was existing already and I don't want to break that if it is used elsewhere as we know scrut does get things wrong sometimes :)

**Comment by laf** [comment]
'trigger ci'

**Comment by laf** [comment]
trigger ci

**Comment by f0o** [comment]
trigger ci

**Comment by tylerapplebaum** [comment]
This looks great! When will it be available in production?

**Comment by laf** [comment]
*cough* @f0o / @paulgear :)

**Comment by laf** [comment]
lol, close and comment is not what I wanted to do :/

**Comment by f0o** [comment]
trigger ci

**Comment by f0o** [comment]
Will merge when this test goes through and it all looks good :)

**Comment by tylerapplebaum** [comment]
Looks great!
image

**Pull request merged**